fdbclient/CommitProxyInterface.h

```diff
@@ -116,18 +116,30 @@ struct ClientDBInfo {
 	    firstCommitProxy; // not serialized, used for commitOnFirstProxy when the commit proxies vector has been shrunk
 	Optional<Value> forward;
 	vector<VersionHistory> history;
+	vector<std::pair<UID, StorageServerInterface>>
```

Comment:
I didn't realize that you're shipping these through the commit proxy interface. This means that we will ship this to every client on every recovery. Not only that, we will ship as many copies as there are commit proxies. I fear this could become a scaling bottleneck. The number of bytes you're adding to the clientDBInfo will now be `number_of_tss * number_of_commit_proxies * (sizeof(UID) * sizeof(serializedSSInterface)`. Does this also mean you broadcast this through clientDBInfo changes every time a TSS fails?

We have ~300 storages in prod, 3 proxies, ~2k clients, and 360 servers where this needs to be broadcasted to. I don't think this will work. Instead I would suggest to change `GetKeyServerLocationsReply` to include TSS.

Reply:
Thanks a ton for reviewing Markus!
Yes, that's correct. Both on recovery and when a TSS enters/leaves the cluster/restarts. I think we discussed this in the first TSS review meeting when you were on PTO.
@sfc-gh-etschannen and I had initially talked about this as a potential scaling issue, but thought it might be fine and had decided to punt on it for now.
One improvement we had identified if it became an issue was to just broadcast a hash/generation of the tss mapping in the ClientDBInfo, and have the client re-read the mapping from the system keyspace only if it changed since the last time.
For propagating it through the GetKeyServerLocationsReply, I can think of several approaches.
You could just include the whole mapping in every reply, which shouldn't be too incrementally larger than the existing reply size. You could also optimize this using the generation/hash type thing, where the reply includes the hash/generation/version of the tss mapping, and the client does a separate thing to re-read the mapping only if it changed.

Reply:
I don't know whether it is fine -- my intuition would say no, but I might very well be wrong. But I fear think testing this will be more work then just sending the tss interfaces with each `GetKeyServerLocationsReply` (that would only increase the size of that reply and the client will cache it properly)...

Maybe I am missing something here, but what is the difficulty and/or drawback of changing `GetKeyServerLocationsReply`? All the logic to maintain this cache on the client is already implemented and tested. TSS will only be assigned during recruitment (so you never have to worry not to know about an existing TSS) and for failed TSS you could simply query for the TSS once every 5 minutes or so.

Reply:
Updating after offline discussion - we are going to merge this as is and attempt this change in a separate PR.